strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactGroupServiceImpl.java

```diff
@@ -0,0 +1,95 @@
+package org.carlspring.strongbox.services.impl;
+
+import org.carlspring.strongbox.data.service.CommonCrudService;
+import org.carlspring.strongbox.domain.ArtifactGroup;
+import org.carlspring.strongbox.services.ArtifactGroupService;
+
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import com.orientechnologies.common.concur.ONeedRetryException;
+import com.orientechnologies.orient.core.record.impl.ODocument;
+import com.orientechnologies.orient.core.sql.query.OSQLSynchQuery;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Przemyslaw Fusik
+ */
+@Service
+@Transactional
+public class ArtifactGroupServiceImpl
+        extends CommonCrudService<ArtifactGroup>
+        implements ArtifactGroupService
+{
+
+    @Override
+    public Class<ArtifactGroup> getEntityClass()
+    {
+        return ArtifactGroup.class;
+    }
+
+    @Override
+    public <T extends ArtifactGroup> T findOneOrCreate(Class<T> type,
+                                                       String name)
+    {
+        Optional<T> optional = tryFind(type, name);
+        if (optional.isPresent())
+        {
+            return optional.get();
+        }
+
+        T artifactGroup = instantiateNew(type, name);
+
+        try
+        {
+            return save(artifactGroup);
+        }
+        catch (ONeedRetryException ex)
+        {
+            optional = tryFind(type, name);
+            if (optional.isPresent())
+            {
+                return optional.get();
+            }
+            throw ex;
+        }
+    }
+
+    private <T extends ArtifactGroup> T instantiateNew(Class<T> type,
+                                                       String name)
+    {
+        T artifactGroup;
+        try
+        {
+            artifactGroup = type.getConstructor(String.class)
```

Comment:
this is not really generic :)

Reply:
yes, it is. Every artifact group has a name. I would make it the only one constructor without orientDb needs to have no-args constructor too.

Reply:
the thing is that you can't say what `ArtifactGroup` implementations will be in the future. And there can be not only `name` needed for some implementations.

Reply:
`name` dropped